### PR TITLE
Restart FTL when reducing the privacy level

### PR DESCRIPTION
**Pull request by DL6ER**
# What does this implement/fix?

Always restart FTL when changing the privacy level

**Related issue or feature (if applicable):** N/A

**Pull request in [docs](https://github.com/pi-hole/docs) with documentation (if applicable):** N/A

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

## Checklist:

- [x] The code change is tested and works locally.
- [x] I based my code and PRs against the repositories `developmental` branch.
- [x] I [signed off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits. Pi-hole enforces the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions
- [x] I [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) all my commits. Pi-hole requires signatures to verify authorship
- [x] I have read the above and my PR is ready for review.

### CODE REVIEW

**Review by yubiuser — COMMENTED** [review]
Mhh, in v5 a restart was only required when lowering the level, not increasing. Did this change? If so, we need to adjust the help text.

> The privacy level may be increased at any time without having to restart the DNS resolver. However, note that the DNS resolver needs to be restarted when lowering the privacy level.

**Comment by DL6ER** [comment]
We needed quite a drastic restructuring of the code to achieve a restarting of FTL based on the difference between an old and a new value but this should now be possible. I did a lot of testing myself but please test this as much as you can, too.

File: src/config/config.c [hunk 1]
```diff
@@ -1490,8 +1493,16 @@ void reread_config(void)
 		   strcmp(conf_copy.webserver.api.pwhash.v.s, config.webserver.api.pwhash.v.s) != 0)
 			delete_all_sessions();
 
+		// Check if privacy level was reduced. If so, we need to restart FTL
+		if(conf_copy.misc.privacylevel.v.privacy_level < config.misc.privacylevel.v.privacy_level)
+		{
+			log_info("Privacy level was reduced, restarting FTL");
```
Comment (yubiuser) [inline src/config/config.c:1499]: This part of the code (logging) is never reached

```
nanopi@nanopi:~$ cat /var/log/pihole/FTL.log | grep Privacy
nanopi@nanopi:~$ 
```

**Review by yubiuser — COMMENTED** [review]
There is quite a huge delay between saving the new settings ans executing the restart.

```
2023-10-20 22:31:41.246 [183281/T183627] INFO: Writing config file
2023-10-20 22:31:49.690 [183281M] INFO: Shutting down... // exit code 22 // jmpret 0
```
Could this be improved?

File: src/config/config.c (hunk 1 above)
Comment (DL6ER) [reply src/config/config.c:1499]: It is:
```
2023-10-21 13:22:53.770 [1057441/T1057444] INFO: Privacy level was reduced, restarting FTL
```

Mind that there are two ways of changing FTL's configuration: *internally* (via the API) or *externally* (via manipulating `pihole.toml` by a third-party `pihole-FTL --config ...` call) and the code we are talking about here is in the *external* path. So if you run
```
pihole-FTL --config misc.privacylevel 0
```
after it had been set to, e.g., `1`, FTL will actually log this as seen above in my comment.

**Comment by DL6ER** [comment]
@yubiuser The latest commit should improve on the observed delay:

```
...
2023-10-21 13:28:49.133 [1059758/T1059765] INFO: Privacy level was reduced, restarting FTL
2023-10-21 13:28:49.147 [1059758/T1059765] INFO: Writing config file
2023-10-21 13:28:49.153 [1059758M] INFO: Shutting down... // exit code 22 // jmpret 0
2023-10-21 13:28:49.451 [1059758M] INFO: Finished final database update
2023-10-21 13:28:49.451 [1059758M] INFO: Waiting for threads to join
2023-10-21 13:28:49.451 [1059758M] INFO: Thread database (0) is idle, terminating it.
2023-10-21 13:28:49.451 [1059758M] INFO: Thread housekeeper (1) is idle, terminating it.
2023-10-21 13:28:49.451 [1059758M] INFO: Thread DNS client (2) is idle, terminating it.
2023-10-21 13:28:49.451 [1059758M] INFO: All threads joined
2023-10-21 13:28:50.116 [1059758M] INFO: ########## FTL terminated after 20s  (code 22)! ##########
2023-10-21 13:28:50.133 [1059758M] INFO: ########## FTL started on ryzen! ##########
2023-10-21 13:28:50.133 [1059758M] INFO: FTL branch: fix/privacy_restart
2023-10-21 13:28:50.133 [1059758M] INFO: FTL version: v5.23-1080-gc30418bd-dirty
2023-10-21 13:28:50.133 [1059758M] INFO: FTL commit: c30418bd-dirty
2023-10-21 13:28:50.133 [1059758M] INFO: FTL date: 2023-10-20 19:32:00 +0200
2023-10-21 13:28:50.133 [1059758M] INFO: FTL user: pihole
...
```

File: src/config/config.c (hunk 1 above)
Comment (yubiuser) [reply src/config/config.c:1499]: I used the webinterface to change the privacy level

**Review by yubiuser — COMMENTED** [review]
Now it restarts super fast. Could you add some (generic) logging when this happens. (Also for restarts triggered internally)

"FTL configuration changed - restart required."

**Comment by DL6ER** [comment]
It is already visible in the `exit code 22` part, but I agree that it'd be made even more obvious. Added in the most recent commit.

**Comment by yubiuser** [comment]
Knowing how the exit code translates is considered expert knowledge.

**Review by yubiuser — APPROVED** [review]
```
2023-10-21 19:43:38.679 [195495/T197250] INFO: Reloading config due to pihole.toml change
2023-10-21 19:43:38.686 [195495/T197250] INFO: Privacy level was reduced, restarting FTL
2023-10-21 19:43:38.711 [195495/T197250] INFO: Writing config file
2023-10-21 19:43:38.724 [195495/T197250] INFO: Restarting FTL due to pihole.toml change
2023-10-21 19:43:38.726 [195495M] INFO: Shutting down... // exit code 22 // jmpret 0
```

```
2023-10-21 19:42:03.423 [195495/T195546] INFO: Writing config file
2023-10-21 19:42:03.439 [195495/T195546] INFO: Restarting FTL due to API config change
2023-10-21 19:42:03.442 [195495M] INFO: Shutting down... // exit code 22 // jmpret 0

```

**Pull request merged**